sisl/geom/nanoribbon.py

```diff
@@ -154,3 +159,533 @@ def zgnr(width, bond=1.42, atoms=None):
+
+_heteroribbon_section = namedtuple("Section", ["W", "L", "shift", "align", "atoms"], defaults=(1, 0, None, None))
+
+
+@set_module("sisl.geom")
+def heteroribbon(sections, bond, atoms, kind="armchair",
+                 invert_first=False, align="bottom",
+                 correct_shifts=False, on_dangling="raise", pbc=True):
+    """Build a nanoribbon consisting of several nanoribbons of different widths
+
+    Parameters
+    ----------
+    sections: array-like of (heteroribbon.Section or tuple or dict)
+        A list of sections to be added to the ribbon.
+
+        Each section is either a `heteroribbon.Section` or an array that will
+        be converted to a `heteroribbon.Section`. Therefore it should contain:
+            W: int
+                The width of the section.
+            L: int, optional
+                The number of units of the section. Note that a "unit" is
+                not a unit cell, but half of it. I.e. a transversal string of
+                atoms.
+
+                If not provided, it defaults to 1.
+            shift: int, optional
+                The shift of this section with respect to the previous one.
+                It can be both positive (upwards shift) or negative (downwards shift).
+
+                If not provided, it defaults to 0.
+            align: {"bottom"/"b", "top"/"t", "center"/"c", "auto"/"a"}
+                Indicates how the section should be aligned with respect to the
+                previous one, by default this will be the `align` argument passed to this
+                function.
+            atoms: Atom, optional
+                Value to pass to the `atoms` argument of `nanoribbon`. If not provided,
+                it defaults to the `atoms` argument passed to this function.
+    bond: float
+        The bond length of the ribbon.
+    atoms: Atom
+        The default atom to use for the sections. For each section, if it contains an `atoms`
+        value, it will be used instead of this one.
+    kind: {'armchair', 'zigzag'}
+        The kind of ribbon to be produced.
+    invert_first: bool, optional
+        Whether the first section should be inverted with respect to the one provided by
+        `sisl.geom.nanoribbon`.
+    align: {"bottom"/"b", "top"/"t", "center"/"c", "auto"/"a"}
+        The default alignment between sections.
+        Sections `align` argument in their `heteroribbon.Section` will have precedence.
+
+        If automatic alignment is requested, sections are aligned:
+         - If both sections are odd: On their center.
+         - If previous section is even: On its open edge (top or bottom)
+         - If previous section is odd and incoming section is even: On the bottom.
+    correct_shifts: bool, optional
+        Whether or not the implementation will correct faulty shifts.
+
+        If ``False``, sections are just shifted (`shift`) number of atoms.
+
+        If ``True``, shifts are quantized in the sense that shifts that produce dangling
```

Comment:
`...that shifts that produce low-coordinated atoms (neighbours < 2) are ignored`?